### DISCUSSION TITLE
How to do reusable components with x-modelable?

### DISCUSSION THREAD

**Discussion by kslstn** [alpinejs/alpine#3584] [1. Help] [answered]
I'm a bit confused about how to do reuse code. Assume we have server-side rendering and a component like in the docs:
```
<div x-data="dropdown">
    <button @click="toggle">...</button>
 
    <div x-show="open">...</div>
</div>
 
<script>
    document.addEventListener('alpine:init', () => {
        Alpine.data('dropdown', () => ({
            open: false,
 
            toggle() {
                this.open = ! this.open
            }
        }))
    })
</script>
```
Now if I add two of those dropdowns, they'd open in sync, because they share the same `x-data` (_Edit: that's not true_). I can circumvent that by adding `x-data` into the HTML template and adding a unique property name for each instance instead of 'open'. 

That works fine but for components that can do a little more, that leads to a lot of duplicate code in the final template; code where only a few property names differ. It also requires that I write JS in an HTML context. With that I loose the syntax highlighting and whatever else I get when editing JS files separately in my current setup.

I considered setting `open` via an argument for `x-data`:
```
x-data="dropdown(open = parentProperty)"
```
But then I'd have to let the component listen to the parent property for changes and still update the parent property with the component's changes. That's basically reinventing two-way binding and quite unwieldy.

Surely someone has some better ideas?

**Comment by SimoTod** [comment]
No, Alpine.data is a factory, every component gets a fresh instance of the data.

**Comment by ekwoka** [comment]
Did you try the first code you show there?

because it would have each component instance be unique.

**Comment by kslstn** [comment]
Ah sorry, apparently there was something in my project that made me believe the first issue was real. I [made a Codepen](https://codepen.io/kslstn/pen/yLRQMyd?editors=1111) that does what I need, but it feels a bit unwieldy to add a watcher in the template and `x-data` and dispatch changes: 

```
document.addEventListener('alpine:init', () => {
	Alpine.data('myComponent', (options) => ({
		myProperty: options.parentProperty,
		parentPropertyName: options.parentPropertyName,
		
		init(){
			this.$watch(this.parentPropertyName, (val)=>{ this.myProperty = val })
			this.$watch('myProperty', (val)=>{ this.$dispatch('change', val) })
		},
	})
)});
```
(There must be a cleaner way to get the parent property name and value into the component, but `myProperty: this[options.parentPropertyName]` becomes undefined?)

**Reply by SimoTod** [reply, marked as answer]
you use x-model+x-modelable: https://codepen.io/SimoTod/pen/QWZJvmp

**Comment by kslstn** [comment]
Turns out my initial and current issue is specific to typing text into text inputs. 

With SimoTod's help I got it mostly working, except that text changes are ignored by Alpine. The odd thing is that by changing the property via  a `@keyup` handler does work. Here's another [Codepen showing that](https://codepen.io/kslstn/pen/RweqLJb?editors=1111).

There's [this issue](https://github.com/alpinejs/alpine/discussions/3081) describing a similar problem, but surely me and the other person aren't the only ones trying to use x-modelable with inputs?

**Reply by SimoTod** [reply]
X-modelable and x-model are meant to be defined on the root of your component next to x-data (no input). Surely it works internally as well but there are gotchas as you saw. Then internally you use x-model pointing to the internal state. Basically what x-modelable does is that it hooks into x-model and triggers a consistent uodate but, in your example, x-model points directly to the external variable so the intetnal one won't update most of the time.

**Reply by kslstn** [reply]
Thanks, I believe what you're saying is consistent with what I see in my [updated Codepen](https://codepen.io/kslstn/pen/RweqLJb?editors=1011) with some different use cases.

To summarize for future reference (please correct me if I'm wrong):
- `x-model` refers to the parent component's property, `x-modelable` connects that to a child's property.
- `x-data`, `x-model` and `x-modelable` need to be set on the root-level element of a component
- That root element can't be an input; inputs need to be wrapped inside an other element.
- If you have an input inside a child that should update a parent's property via `x-modelable`, it needs an `x-model` with the same property name as the child's `x-modelable` property.

**Reply by vjbento** [reply]
Thanks @kslstn!!!

**Reply by ekwoka** [reply]
the x-model and x-modelable do not need to have matching expressions.

It's basically x-modelable defines the child components inner state name, and the x-model defines the parents state name, and this entangles those values together.
